### PR TITLE
Tpetra:  Added extra CMake logic for setting Tpetra_ASSUME_CUDA_AWARE_MPI

### PR DESCRIPTION
**Pull request by kddevin**
For issue #3405 
These changes add logic to CMake's configuration, increasing the number of locations in which ompi_info is found.  The changes allow ompi_info to be found for ATDM builds on white.
ATDM builds did not set MPI_CXX_COMPILER; instead, they set CMAKE_CXX_COMPILER.  The new logic looks for ompi_info in the same directory as CMAKE_CXX_COMPILER.

I also added a test program, cudaMpi_CanaryTest.cpp.  It exercises MPI with CUDA directly, without any Kokkos or Trilinos data structures.  If this test fails with a given MPI installation (e.g., OpenMPI3 on white), the problem is likely with MPI, not with Tpetra.  Is there a better place in Trilinos for such tests, rather than in the packages?  Do we need a "stableSystemPackage" that includes tests like these, which have no dependence on Trilinos yet flag conditions that will cause Trilinos to fail?

Note that once this change is merged, ATDM tests using OpenMPI3 on white will fail unless they explicitly set Tpetra_ASSUME_CUDA_AWARE_MPI=OFF.  _**Since OpenMPI3 is known to be bad on white, users should probably use OpenMPI2.**_

@bartlettroscoe @fryeguy52  To save you headaches, I will not set AT:AUTOMERGE for this PR; feel free to merge when you are ready.

@trilinos/tpetra @bartlettroscoe @fryeguy52 

## How Has This Been Tested?

Implemented and exercised on white.
Exercised on linux workstation.

## Checklist

- [ x] My commit messages mention the appropriate GitHub issue numbers.
- [ x] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [ x] I have added tests to cover my changes.
- [x ] All new and existing tests passed.
- [x ] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

File: packages/tpetra/CMakeLists.txt [hunk 1]
```diff
@@ -172,6 +172,21 @@ ELSE ()
       ENDIF ()
     ENDIF ()
 
+    IF ((NOT Tpetra_FOUND_OMPI_INFO_EXECUTABLE) AND (DEFINED MPI_USE_COMPILER_WRAPPERS) AND MPI_USE_COMPILER_WRAPPERS AND (DEFINED CMAKE_CXX_COMPILER))
```
Comment (mhoemmen) [inline packages/tpetra/CMakeLists.txt]: This looks like an exact duplicate of the logic above.  Am I missing something, or does CMake weirdly need this to work?

Comment (kddevin) [reply packages/tpetra/CMakeLists.txt]: The check here is for CMAKE_CXX_COMPILER rather than MPI_CXX_COMPILER; that's the only difference.

Comment (mhoemmen) [reply packages/tpetra/CMakeLists.txt]: @kddevin Thanks for clarifying :D

File: packages/tpetra/core/test/Comm/cudaMpi_CanaryTest.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+// with Trilinos.  If this test fails, there is likely something wrong
+// with the MPI installation on the CUDA platform.
+// See Trilinos github issue #3405 and related test cudaAwareMpi.cpp
+// in this directory.
+
+#include <iostream>
+#include <cstdio>
+#include <mpi.h>
+
+int main(int narg, char **arg)
```
Comment (bartlettroscoe) [inline packages/tpetra/core/test/Comm/cudaMpi_CanaryTest.cpp]: @kddevin ,

This looks like a great test to have in the test suite.   This is basically an acceptance test for what Tpetra needs out of MPI if the CUDA-aware option is on.

Thanks for adding this!

**Comment by bartlettroscoe** [comment]
@kddevin and @mhoemmen, my experimentation in #2674 suggest that enabling this option might actually make more tests fail.  Therefore, if this gets merged to the 'develop' branch, will there be people resources to debug and address the new failures?  We could migrate this risk some by running the entire ATDM Trilinos test suite on say 'waterman' to see what impact this has.

**Comment by mhoemmen** [comment]
@bartlettroscoe wrote:
> Therefore, if this gets merged to the 'develop' branch, will there be people resources to debug and address the new failures?

Trilinos developers need to care and are paid to care about MPI + CUDA issues.  I want to see those failures so we can fix them, whether that means fixing the MPI configuration or fixing Trilinos itself.  So yes, there will be people resources :-)

**Comment by bartlettroscoe** [comment]
> Trilinos developers need to care and are paid to care about MPI + CUDA issues. I want to see those failures so we can fix them, whether that means fixing the MPI configuration or fixing Trilinos itself. So yes, there will be people resources :-)

@mhoemmen, okay then, can we at least do a full run of the ATDM Trilinos test suite on 'wateman' before we merge this in?

**Comment by mhoemmen** [comment]
@bartlettroscoe I'm OK with that.

**Comment by kddevin** [comment]
Hi @bartlettroscoe  .  Just curious:  have you had time to revisit this now that there are better devpacks on white?  Thanks.

**Comment by bartlettroscoe** [comment]
> Hi @bartlettroscoe . Just curious: have you had time to revisit this now that there are better devpacks on white? Thanks.

@kddevin, the env on 'white' and 'ride' was updated and is fine now.  We should do some testing on 'white'/'ride' and 'waterman' to see what impact turning this option on has on those machines.  I can fire off some builds and see how it goes.  Otherwise, if this causes some new failures, does the Tpetra team have time to pretty quickly address the failures?

**Comment by kddevin** [comment]
Thanks for offering to give this a try, @bartlettroscoe 
Let's see how many and which problems it causes; then we can figure out a schedule for fixing them.  We definitely want this capability to work, so fixing problems will be a priority.

**Comment by bartlettroscoe** [comment]
> Let's see how many and which problems it causes; then we can figure out a schedule for fixing them. We definitely want this capability to work, so fixing problems will be a priority.

@kddevin, do you mind if I rebase this branch on top of 'develop' so this tests against the current version of Trilinos?

We can also try back this out if it causes problem problems that I can reasonably test for locally.

**Comment by kddevin** [comment]
Whatever is easiest for you is fine with me, @bartlettroscoe 
Thanks.

**Comment by bartlettroscoe** [comment]
@trilinos/framework,

FYI: The autotester put the label "AT: STALE" on this PR as the last test was run on 9/26/2018.  Yet, the "Merge pull request" button is active and I could merge this PR!  Is this a defect in the autotester?

**Comment by bartlettroscoe** [comment]
@kddevin and @mhoemmen,

FYI: I have rebased this branch on top of 'develop' (and forced pushed) and I am currently testing this on 'waterman' by running:

```
$ bsub -x -Is -n 20 ./checkin-test-atdm.sh \
   cuda-9.2-debug-Power9-Volta70 \
   cuda-9.2-opt-Power9-Volta70 \
   cuda-9.2-release-debug-Power9-Volta70 \
   --enable-packages=Tpetra --enable-fwd-packages --local-do-all 
```

I will report back here what I see.

**Comment by jwillenbring** [comment]
@bartlettroscoe 

> FYI: The autotester put the label "AT: STALE" on this PR as the last test was run on 9/26/2018. Yet, the "Merge pull request" button is active and I could merge this PR! Is this a defect in the autotester?

I am not sure if something changed already, but the merge button is not active for me.

**Comment by bartlettroscoe** [comment]
> I am not sure if something changed already, but the merge button is not active for me.

@jwillenbring, sorry, I messed this up when I forced pushed a rebased branch just now.  I promise you, the merge button was green.

**Comment by bartlettroscoe** [comment]
@kddevin and @mhoemmen,

Yesterday I rebased the branch on top of 'develop' and then on 'waterman' I ran:

```
$ bsub -x -Is -n 20 ./checkin-test-atdm.sh \
   cuda-9.2-debug-Power9-Volta70 \
   cuda-9.2-opt-Power9-Volta70 \
   cuda-9.2-release-debug-Power9-Volta70 \
   --enable-packages=Tpetra --enable-fwd-packages --local-do-all 
```

This gave the result shown in detail below.  In summary, in all of the builds with optimized compiler options, it gave two new failing tests:

* 1164 - PanzerAdaptersSTK_PoissonInterfaceExample_3d_MPI_4 (Failed)
* 1166 - PanzerAdaptersSTK_PoissonInterfaceExample_3d_MPI_4-ConvTest (Failed)

I submitted the results for for the build `cuda-9.2-opt-Power9-Volta70` to CDash by doing:

```
$ cd cuda-9.2-opt-Power9-Volta70/

$ . load-env.sh 
Hostname 'waterman11' matches known ATDM host 'waterman' and system 'waterman'
ATDM_CONFIG_TRILNOS_DIR = /home/rabartl/Trilinos.base/Trilinos
Setting default compiler and build options for ATDM_CONFIG_JOB_NAME='cuda-9.2-opt-Power9-Volta70'
Using waterman compiler stack CUDA-9.2 to build RELEASE code with Kokkos node type CUDA and KOKKOS_ARCH=Power9,Volta70

$  bsub -x -Is -n 20 \
   env CTEST_DO_CONFIGURE=OFF CTEST_DO_BUILD=OFF CTEST_PARALLEL_LEVEL=8 \
   make dashboard
```

This submitted results to:

* https://testing.sandia.gov/cdash-dev-view/index.php?project=Trilinos&parentid=4178723

with the failing tests link:

* https://testing.sandia.gov/cdash-dev-view/viewTest.php?onlyfailed&buildid=4178730

Looks like a segfault:

```
[waterman1:13668] *** Process received signal ***
[waterman1:13668] Signal: Aborted (6)
[waterman1:13668] Signal code:  (-6)
```

I think it would be a good idea to fix these failures before doing testing on other platforms before we merge this.

Seem reasonble?

<details>
<summary><b>DETAILED TEST RESULTS</b> (click to expand)</summary>

```
Enabled Packages: Tpetra

Build test results:
-------------------
0) MPI_RELEASE_DEBUG_SHARED_PT_OPENMP => Test case MPI_RELEASE_DEBUG_SHARED_PT_OPENMP was not run! => Does not affect push readiness! (-1.00 min)
1) cuda-9.2-debug-Power9-Volta70 => passed: passed=1159,notpassed=0 (106.95 min)
2) cuda-9.2-opt-Power9-Volta70 => FAILED: passed=1181,notpassed=2 => Not ready to push! (95.27 min)
3) cuda-9.2-release-debug-Power9-Volta70 => FAILED: passed=1180,notpassed=2 => Not ready to push! (123.50 min)

1) cuda-9.2-debug-Power9-Volta70 Results:
-----------------------------------------

  passed: Trilinos/cuda-9.2-debug-Power9-Volta70: passed=1159,notpassed=0
  
  Wed Nov 14 15:04:59 MST 2018
  
  Enabled Packages: Tpetra
  Enabled all Forward Packages
  Hostname: waterman4
  Source Dir: /home/rabartl/Trilinos.base/Trilinos/cmake/tribits/ci_support/../../..
  Build Dir: /home/rabartl/Trilinos.base/BUILDS/WATERMAN/CHECKIN/cuda-9.2-debug-Power9-Volta70
  
  CMake Cache Varibles: -GNinja -DTrilinos_TRIBITS_DIR:PATH=/home/rabartl/Trilinos.base/Trilinos/cmake/tribits -DTrilinos_ENABLE_TESTS:BOOL=ON -DTrilinos_TEST_CATEGORIES:STRING=NIGHTLY -DTrilinos_ALLOW_NO_PACKAGES:BOOL=OFF -DDART_TESTING_TIMEOUT:STRING=600.0 -GNinja -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake -DTrilinos_TRACE_ADD_TEST=ON -DTrilinos_ENABLE_Tpetra:BOOL=ON -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON -DTrilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES:BOOL=ON
  Make Options: -j 64
  CTest Options: -j 8
  
  Pull: Not Performed
  Configure: Passed (2.79 min)
  Build: Passed (58.62 min)
  Test: Passed (45.53 min)
  
  100% tests passed, 0 tests failed out of 1159
  
  Subproject Time Summary:
  Amesos2        =  92.25 sec*proc (7 tests)
  Anasazi        = 1121.86 sec*proc (74 tests)
  Belos          = 1121.96 sec*proc (96 tests)
  Ifpack2        = 1062.28 sec*proc (45 tests)
  MueLu          = 6033.81 sec*proc (99 tests)
  NOX            = 561.46 sec*proc (105 tests)
  Panzer         = 2541.90 sec*proc (136 tests)
  Piro           =  33.94 sec*proc (12 tests)
  Rythmos        = 129.22 sec*proc (83 tests)
  Stratimikos    =  82.54 sec*proc (39 tests)
  Teko           = 1309.48 sec*proc (19 tests)
  Tempus         = 1344.90 sec*proc (54 tests)
  Thyra          = 290.45 sec*proc (81 tests)
  Tpetra         = 3765.37 sec*proc (190 tests)
  Xpetra         = 578.04 sec*proc (18 tests)
  Zoltan2        = 1419.62 sec*proc (101 tests)
  
  Total Test time (real) = 2731.93 sec
  
  Total time for cuda-9.2-debug-Power9-Volta70 = 106.95 min


2) cuda-9.2-opt-Power9-Volta70 Results:
---------------------------------------

  FAILED: Trilinos/cuda-9.2-opt-Power9-Volta70: passed=1181,notpassed=2
  
  Wed Nov 14 16:40:22 MST 2018
  
  Enabled Packages: Tpetra
  Enabled all Forward Packages
  Hostname: waterman4
  Source Dir: /home/rabartl/Trilinos.base/Trilinos/cmake/tribits/ci_support/../../..
  Build Dir: /home/rabartl/Trilinos.base/BUILDS/WATERMAN/CHECKIN/cuda-9.2-opt-Power9-Volta70
  
  CMake Cache Varibles: -GNinja -DTrilinos_TRIBITS_DIR:PATH=/home/rabartl/Trilinos.base/Trilinos/cmake/tribits -DTrilinos_ENABLE_TESTS:BOOL=ON -DTrilinos_TEST_CATEGORIES:STRING=NIGHTLY -DTrilinos_ALLOW_NO_PACKAGES:BOOL=OFF -DDART_TESTING_TIMEOUT:STRING=600.0 -GNinja -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake -DTrilinos_TRACE_ADD_TEST=ON -DTrilinos_ENABLE_Tpetra:BOOL=ON -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON -DTrilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES:BOOL=ON
  Make Options: -j 64
  CTest Options: -j 8
  
  Pull: Not Performed
  Configure: Passed (2.78 min)
  Build: Passed (41.95 min)
  Test: FAILED (50.54 min)
  
  99% tests passed, 2 tests failed out of 1183
  
  Subproject Time Summary:
  Amesos2        = 139.12 sec*proc (8 tests)
  Anasazi        = 1094.59 sec*proc (74 tests)
  Belos          = 1195.62 sec*proc (96 tests)
  Ifpack2        = 806.88 sec*proc (45 tests)
  MueLu          = 6389.68 sec*proc (99 tests)
  NOX            = 424.02 sec*proc (105 tests)
  Panzer         = 6117.85 sec*proc (159 tests)
  Piro           =  32.53 sec*proc (12 tests)
  Rythmos        =  83.72 sec*proc (83 tests)
  Stratimikos    =  46.04 sec*proc (39 tests)
  Teko           = 1144.94 sec*proc (19 tests)
  Tempus         = 567.93 sec*proc (54 tests)
  Thyra          = 316.30 sec*proc (81 tests)
  Tpetra         = 3610.77 sec*proc (190 tests)
  Xpetra         = 542.92 sec*proc (18 tests)
  Zoltan2        = 1394.49 sec*proc (101 tests)
  
  Total Test time (real) = 3032.10 sec
  
  The following tests FAILED:
  	1164 - PanzerAdaptersSTK_PoissonInterfaceExample_3d_MPI_4 (Failed)
  	1166 - PanzerAdaptersSTK_PoissonInterfaceExample_3d_MPI_4-ConvTest (Failed)
  Errors while running CTest
  
  Total time for cuda-9.2-opt-Power9-Volta70 = 95.27 min


3) cuda-9.2-release-debug-Power9-Volta70 Results:
-------------------------------------------------

  FAILED: Trilinos/cuda-9.2-release-debug-Power9-Volta70: passed=1180,notpassed=2
  
  Wed Nov 14 18:43:58 MST 2018
  
  Enabled Packages: Tpetra
  Enabled all Forward Packages
  Hostname: waterman4
  Source Dir: /home/rabartl/Trilinos.base/Trilinos/cmake/tribits/ci_support/../../..
  Build Dir: /home/rabartl/Trilinos.base/BUILDS/WATERMAN/CHECKIN/cuda-9.2-release-debug-Power9-Volta70
  
  CMake Cache Varibles: -GNinja -DTrilinos_TRIBITS_DIR:PATH=/home/rabartl/Trilinos.base/Trilinos/cmake/tribits -DTrilinos_ENABLE_TESTS:BOOL=ON -DTrilinos_TEST_CATEGORIES:STRING=NIGHTLY -DTrilinos_ALLOW_NO_PACKAGES:BOOL=OFF -DDART_TESTING_TIMEOUT:STRING=600.0 -GNinja -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake -DTrilinos_TRACE_ADD_TEST=ON -DTrilinos_ENABLE_Tpetra:BOOL=ON -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON -DTrilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES:BOOL=ON
  Make Options: -j 64
  CTest Options: -j 8
  
  Pull: Not Performed
  Configure: Passed (2.80 min)
  Build: Passed (69.40 min)
  Test: FAILED (51.29 min)
  
  99% tests passed, 2 tests failed out of 1182
  
  Subproject Time Summary:
  Amesos2        = 128.71 sec*proc (7 tests)
  Anasazi        = 1040.68 sec*proc (74 tests)
  Belos          = 1132.81 sec*proc (96 tests)
  Ifpack2        = 950.22 sec*proc (45 tests)
  MueLu          = 6092.51 sec*proc (99 tests)
  NOX            = 483.36 sec*proc (105 tests)
  Panzer         = 7171.49 sec*proc (159 tests)
  Piro           =  66.26 sec*proc (12 tests)
  Rythmos        = 102.09 sec*proc (83 tests)
  Stratimikos    =  44.53 sec*proc (39 tests)
  Teko           = 1142.36 sec*proc (19 tests)
  Tempus         = 589.88 sec*proc (54 tests)
  Thyra          = 292.23 sec*proc (81 tests)
  Tpetra         = 3138.61 sec*proc (190 tests)
  Xpetra         = 548.97 sec*proc (18 tests)
  Zoltan2        = 1454.93 sec*proc (101 tests)
  
  Total Test time (real) = 3077.44 sec
  
  The following tests FAILED:
  	1163 - PanzerAdaptersSTK_PoissonInterfaceExample_3d_MPI_4 (Failed)
  	1165 - PanzerAdaptersSTK_PoissonInterfaceExample_3d_MPI_4-ConvTest (Failed)
  Errors while running CTest
  
  Total time for cuda-9.2-release-debug-Power9-Volta70 = 123.50 min
```

</details>

**Comment by kddevin** [comment]
@bartlettroscoe  Thank you for running these tests.  I am pleased that so few failed. 

Let's pull the @trilinos/panzer folks in to look at these.

@rppawlo @trilinos/panzer This PR is enabling CUDA-aware MPI in ATDM builds.  Two Panzer tests fail in tests by @bartlettroscoe on waterman.  Please see the failing tests link above for details.  
For both failures, the error message before the segfault is 
"Panzer Warning: No global indexer assigned to a workset container. Orientation of the basis for edge basis functions cannot be applied, if those basis functions are used, there will be problems!"
Thanks for taking a look.

**Comment by rppawlo** [comment]
I've narrowed this down to a Tpetra export call. Going to need @mhoemmen 's help on this one.

**Comment by mhoemmen** [comment]
We're scheduled to debug together this afternoon.  @rppawlo try setting the environment variable `TPETRA_VERBOSE` to 1 before running, to help narrow this down.

**Comment by rppawlo** [comment]
> try setting the environment variable `TPETRA_VERBOSE` to 1 before running, to help narrow this down.

@mhoemmen - I've got all that data saved and ready for you to look at.

**Comment by rppawlo** [comment]
This is blocked by #3968

**Comment by rppawlo** [comment]
@bartlettroscoe - everything is passing on waterman now with CUDA_AWARE_MPI enabled. You should be able to proceed with this.

**Comment by bartlettroscoe** [comment]
>  everything is passing on waterman now with CUDA_AWARE_MPI enabled. You should be able to proceed with this.

@rppawlo, I will rebase this branch and then run some experimental CUDA 9.2 builds on 'ride' and 'waterman' testing from Tpetra on down just to make sure.  (With the removal of CUDA 8.0 and 9.0 builds on 'hansen'/'shiller', these are the only two platforms with CUDA builds left.)  If those show no new failing tests for all the defined CUDA builds, I will go ahead and merge.

**Comment by bartlettroscoe** [comment]
FYI: I am in the process of running cuda-9.2 builds on 'ride' and 'waterman' but I removed the `AT: WIP` label so that the Trilinos PR tester can run the builds.  I will report back the results of those tests once they are complete.  If things look good after that, then we can click the "Merge" button.

**Comment by rppawlo** [comment]
I used the Trilinos ATDM build. Details are below. I only ran the Panzer tests but they all passed - including the expensive ones.

```
source ~/Trilinos/cmake/std/atdm/load-env.sh cuda-dbg
```

```
-bash-4.2$ more build_trilinos_dbg.sh
rm -rf CMakeCache.txt CMakeFiles
cmake \
-GNinja \
-DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
-D Tpetra_ENABLE_TESTS=OFF \
-DAmesos2_ENABLE_Epetra=OFF \
-DTrilinos_ENABLE_Panzer=ON \
-DPanzer_ENABLE_TESTS=ON \
-DPanzer_ENABLE_EXAMPLES=ON \
-DPanzer_ADD_EXPENSIVE_CUDA_TESTS=ON \
-DTpetra_INST_INT_INT=ON \
-DTpetra_ASSUME_CUDA_AWARE_MPI=ON \
-DTrilinos_ENABLE_Phalanx=ON \
-DMueLu_ENABLE_Kokkos_Refactor_Use_By_Default=ON \
-DPhalanx_ENABLE_TESTS=ON \
-DSacado_ENABLE_HIERARCHICAL_DFAD=ON \
-DTrilinos_ENABLE_SEACASExodiff=ON \
-DTrilinos_ENABLE_SEACASEpu=ON \
-DTrilinos_ENABLE_SEACASNemspread=ON \
-DTrilinos_ENABLE_SEACASNemslice=ON \
-D Trilinos_ENABLE_SEACASAprepro:BOOL=ON \
-DKokkos_ENABLE_Cuda_Relocatable_Device_Code=OFF \
-DTrilinos_ENABLE_DEBUG=ON \
-DTrilinos_ENABLE_DEBUG_SYMBOLS:BOOL=ON \
-DTeuchos_ENABLE_DEBUG_RCP_NODE_TRACING=OFF \
~/Trilinos
```

**Comment by rppawlo** [comment]
I used the Trilinos ATDM build. Details are below. I only ran the Panzer tests but they all passed - including the expensive ones.

```
source ~/Trilinos/cmake/std/atdm/load-env.sh cuda-dbg
```

```
-bash-4.2$ more build_trilinos_dbg.sh
rm -rf CMakeCache.txt CMakeFiles
cmake \
-GNinja \
-DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
-D Tpetra_ENABLE_TESTS=OFF \
-DAmesos2_ENABLE_Epetra=OFF \
-DTrilinos_ENABLE_Panzer=ON \
-DPanzer_ENABLE_TESTS=ON \
-DPanzer_ENABLE_EXAMPLES=ON \
-DPanzer_ADD_EXPENSIVE_CUDA_TESTS=ON \
-DTpetra_INST_INT_INT=ON \
-DTpetra_ASSUME_CUDA_AWARE_MPI=ON \
-DTrilinos_ENABLE_Phalanx=ON \
-DMueLu_ENABLE_Kokkos_Refactor_Use_By_Default=ON \
-DPhalanx_ENABLE_TESTS=ON \
-DSacado_ENABLE_HIERARCHICAL_DFAD=ON \
-DTrilinos_ENABLE_SEACASExodiff=ON \
-DTrilinos_ENABLE_SEACASEpu=ON \
-DTrilinos_ENABLE_SEACASNemspread=ON \
-DTrilinos_ENABLE_SEACASNemslice=ON \
-D Trilinos_ENABLE_SEACASAprepro:BOOL=ON \
-DKokkos_ENABLE_Cuda_Relocatable_Device_Code=OFF \
-DTrilinos_ENABLE_DEBUG=ON \
-DTrilinos_ENABLE_DEBUG_SYMBOLS:BOOL=ON \
-DTeuchos_ENABLE_DEBUG_RCP_NODE_TRACING=OFF \
~/Trilinos
```

File: packages/tpetra/core/test/Comm/cudaMpi_CanaryTest.cpp (hunk 2 above)
Comment (mhoemmen) [reply packages/tpetra/core/test/Comm/cudaMpi_CanaryTest.cpp]: In the future, we could enable building this test whether or not it's known at configure time that MPI is CUDA aware, and use the environment variable via Tpetra::Details::Behavior to control its behavior.

**Comment by bartlettroscoe** [comment]
@mhoemmen, you jumped the gun on merging this since we did not get back results from experimental testing on 'ride' just yet.  However, as luck would have it, all of the cuda-9.2 builds on 'ride' 100% passed.  See details and evidence below.

NOTE: We got no testing on 'waterman' because the queue is swamped (see below)..

----

**DETAILED TESTING NOTES:**


I rebased the branch `issue3405` on top of Trilinos 'develop' as of commit:

```
commit aa106410144ac5b4f90aff56df71135e2adf833c
Merge: ab25184 d4bda03
Author: trilinos-autotester <trilinos-autotester@trilinos.org>
Date:   Thu Dec 6 06:18:07 2018 -0700

    Merge Pull Request #4005 from rppawlo/Trilinos/phalanx-default-option
    
    Automatically Merged using Trilinos Pull Request AutoTester
    PR Title: Phalanx default option for multiple evaluator field registration
    PR Author: rppawlo
```

As per the instructions for the new script [ctest-s-local-test-driver.sh](https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md#ctest-s-local-test-driversh) and am running all of the CUDA 9.2 builds on 'ride' with:

```
$ /home/rabartl/Trilinos.base/BUILDS/RIDE/JENKINS_DRIVER/

$ env Trilinos_PACKAGES=Tpetra Trilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES=ON \
  ./ctest-s-local-test-driver.sh cuda-9.2-debug cuda-9.2-opt cuda-9.2-release-debug


***
*** ./ctest-s-local-test-driver.sh  cuda-9.2-debug cuda-9.2-opt cuda-9.2-release-debug
***

ATDM_TRILINOS_DIR = '/home/rabartl/Trilinos.base/Trilinos'

Load some env to get python, cmake, etc ...

Hostname 'ride6' matches known ATDM host 'ride' and system 'ride'
Setting compiler and build options for buld name 'default'
Using white/ride compiler stack GNU to build DEBUG code with Kokkos node type SERIAL and KOKKOS_ARCH=Power8

Running builds: cuda-9.2-debug cuda-9.2-opt cuda-9.2-release-debug

Running Jenkins driver Trilinos-atdm-white-ride-cuda-9.2-debug.sh ...


real    115m28.675s
user    0m0.781s
sys     0m0.282s

Running Jenkins driver Trilinos-atdm-white-ride-cuda-9.2-opt.sh ...


real    101m23.205s
user    0m0.709s
sys     0m0.373s

Running Jenkins driver Trilinos-atdm-white-ride-cuda-9.2-release-debug.sh ...


real    133m22.519s
user    0m0.860s
sys     0m0.232s
```

The configure output showed:

```
$ grep "Tests, Examples" Trilinos-atdm-white-ride-cuda-9.2-debug/SRC_AND_BUILD/BUILD/Testing/Temporary/LastConfigure_20181206-1344.log 
Processing enabled package: Tpetra (Classic, Core, Tests, Examples)
Processing enabled package: Thyra (Core, EpetraAdapters, EpetraExtAdapters, TpetraAdapters, Tests, Examples)
Processing enabled package: Xpetra (Libs, Tests, Examples)
Processing enabled package: Zoltan2 (Libs, Tests, Examples)
Processing enabled package: Belos (Libs, Tests, Examples)
Processing enabled package: Amesos2 (Libs, Tests, Examples)
Processing enabled package: Anasazi (Libs, Tests, Examples)
Processing enabled package: Ifpack2 (Libs, Tests, Examples)
Processing enabled package: Stratimikos (Libs, Tests, Examples)
Processing enabled package: Teko (Libs, Tests, Examples)
Processing enabled package: NOX (Libs, Tests, Examples)
Processing enabled package: MueLu (Libs, Tests, Examples)
Processing enabled package: Rythmos (Libs, Tests, Examples)
Processing enabled package: Tempus (Libs, Tests, Examples)
Processing enabled package: Piro (Libs, Tests, Examples)
Processing enabled package: Panzer (Core, DofMgr, DiscFE, AdaptersSTK, AdaptersIOSS, MiniEM, Tests, Examples)
```

The CMake cache file showed:

```
$ grep ASSUME_CUDA_AWARE_MPI Trilinos-atdm-white-ride-cuda-9.2-debug/SRC_AND_BUILD/BUILD/CMakeCache.txt 
Tpetra_ASSUME_CUDA_AWARE_MPI:BOOL=ON
TPETRA_ASSUME_CUDA_AWARE_MPI:INTERNAL=ON
```

The `TpetraCore_config.h` file showed:

```
$ grep ASSUME_CUDA_AWARE_MPI Trilinos-atdm-white-ride-cuda-9.2-debug/SRC_AND_BUILD/BUILD/packages/tpetra/core/src/TpetraCore_config.h 
#define TPETRA_ASSUME_CUDA_AWARE_MPI
```

So it looks like CUDA-aware MPI support in Tpetra is being correctly enabled and it looks like we are testing from Tpetra on down.

The results on 'ride' are shown at:

* https://testing.sandia.gov/cdash-dev-view/index.php?project=Trilinos&date=2018-12-06&filtercount=2&showfilters=1&filtercombine=and&field1=groupname&compare1=61&value1=Experimental&field2=buildname&compare2=65&value2=Trilinos-atdm-white-ride-cuda-9.2-

The all ran and submitted build and test results (more evidence that jenkins-srn.sandia.gov is broken, see https://sems-atlassian-son.sandia.gov/jira/browse/TRIL-237).

The builds on 'waterman' are not going to run because the queue is completely overloadded with days worth of work (see https://sems-atlassian-son.sandia.gov/jira/browse/TRIL-239).

**Comment by mhoemmen** [comment]
oops, sorry about that!  I was looking at Karen's test and thinking it would be great to have :D

**Pull request merged**